File: iosgl/iosgl/bt/material/MaterialManager.m
//
//  MaterialManager.m
//  iosgl
//
//  Created by zhao on 8/12/2019.
//  Copyright © 2019 zhao. All rights reserved.
//
#import <GLKit/GLKit.h>
#import "TextureRes.h"
#import "Material.h"
#import "TextureLoad.h"
#import "Scene_data.h"
#import "LoadManager.h"
#import "MaterialParam.h"
#import "MaterialLoad.h"
#import "TextureManager.h"
#import "ProgrmaManager.h"
#import "DynamicTexItem.h"
#import "MaterialManager.h"
static MaterialManager *instance = nil;

//private _loadDic: Object;
//private _resDic: Object;
//private _regDic: Object;
@interface MaterialManager()
@property (nonatomic, strong)  NSMutableDictionary* loadDic;
@property (nonatomic, strong)  NSMutableDictionary* resDic;
@property (nonatomic, strong)  NSMutableDictionary* regDic;
@end

@implementation MaterialManager
+ (instancetype)default{
    if (instance == nil) {
        instance = [[MaterialManager alloc] init];
    }
    return instance;
}
- (instancetype)init
{
    self = [super init];
    if (self) {
        self.dic=[[NSMutableDictionary alloc]init];
        _loadDic=[[NSMutableDictionary alloc]init];
        _resDic=[[NSMutableDictionary alloc]init];
        _regDic=[[NSMutableDictionary alloc]init];
    }
    return self;
}
-(TextureRes *) getMaterialByUrl:(NSString*)urlStr;
{
    TextureRes *textureRes=[[TextureRes alloc]init];
  
    textureRes.textTureLuint=[self createTextureWithImage:[UIImage imageNamed:urlStr]];
    return textureRes;
}

- (GLuint)createTextureWithImage:(UIImage *)image;
{
    // 将 UIImage 转换为 CGImageRef
    CGImageRef cgImageRef = [image CGImage];
    GLuint width = (GLuint)CGImageGetWidth(cgImageRef);
    GLuint height = (GLuint)CGImageGetHeight(cgImageRef);
    CGRect rect = CGRectMake(0, 0, width, height);
    
    // 绘制图片
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    void *imageData = malloc(width * height * 4);
    CGContextRef context = CGBitmapContextCreate(imageData, width, height, 8, width * 4, colorSpace, kCGImageAlphaPremultipliedLast | kCGBitmapByteOrder32Big);
    CGContextTranslateCTM(context, 0, height);
    CGContextScaleCTM(context, 1.0f, -1.0f);
    CGColorSpaceRelease(colorSpace);
    CGContextClearRect(context, rect);
    CGContextDrawImage(context, rect, cgImageRef);

    // 生成纹理
    GLuint textureID;
    glGenTextures(1, &textureID);
    glBindTexture(GL_TEXTURE_2D, textureID);
    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, imageData); // 将图片数据写入纹理缓存
    
    // 设置如何把纹素映射成像素
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
    
    // 解绑
    glBindTexture(GL_TEXTURE_2D, 0);
    
    // 释放内存
    CGContextRelease(context);
    free(imageData);
    
    return textureID;
}
-(void)addResByte:(NSString*)url dataByte:(ByteArray*)dataByte;
{
    
    if(! self.dic[url]&&! self.resDic[url]){
        self.resDic[url]=dataByte;
    }
 
}
//  public getMaterialByte($url: string, $fun: Function, $info: Object = null, $autoReg: boolean = false, $regName: string = null, $shader3DCls: any = null): void {
-(void)getMaterialByte:(NSString*)url fun:(SuccessMaterial)fun info:(NSDictionary*)info autoReg:(BOOL)autoReg regName:(NSString*)regName shader3DCls:(NSObject*)shader3DCls;
{
 
    if(self.dic[url]){
        fun(self.dic[url]);
        return;
    }
    MaterialLoad* materialLoad= [[MaterialLoad alloc]init:fun info:info url:url autoReg:autoReg regName:regName shader: shader3DCls];
    if ( _loadDic[url]) {
        NSMutableArray<MaterialLoad*>* ary   =  _loadDic[url];
        [ary addObject:materialLoad];
        return;
    }
    _loadDic[url] = [[NSMutableArray alloc]init];
    [_loadDic[url] addObject:materialLoad];
    if (_resDic[url]) {
        [self meshByteMaterialByt:self.resDic[url] info:materialLoad];
        [_resDic removeObjectForKey:url];
        
    }else{
        [[LoadManager default] load:url type:LoadManager.BYTE_TYPE fun:^(NSObject * _Nonnull value) {

             ByteArray* byte=[[ByteArray alloc]init:[[NSData alloc] initWithContentsOfFile:((NSDictionary*)value)[@"data"]]];
    
            [self meshByteMaterialByt:byte info:(MaterialLoad*)((NSDictionary*)value)[@"info"]];
            
            
          //  this.loadMaterialByteCom($data, _info)
        } info:materialLoad progressFun:^(int pronum) {
            
        }];
    }
 
}
 
-(void)meshByteMaterialByt:(ByteArray*)byte info:(MaterialLoad*)info;
{
    Material* material=[[Material alloc]init];
    [material setByteData:byte];
    material.url = info.url;
    [self loadMaterial:material];
    if (info.autoReg) {
        material.shader=  [[ProgrmaManager default] getMaterialProgram:info.regName shaderCls:info.shader3D material:material paramAry:nil parmaByFragmet:true];
    }
    NSArray<TextureLoad*>* ary = self.loadDic[info.url];
    for (int i = 0; i < ary.count; i++) {
        if(ary[i].info){
              ary[i].fun(@{@"material":material,@"info":ary[i].info });
        }else{
              ary[i].fun(material);
        }
    }
    [self.loadDic removeObjectForKey:info.url];
    self.dic[info.url] = material;
    
}

-(void)loadDynamicTexUtil:(MaterialParam*)material;
{
    
    NSMutableArray<DynamicTexListVo*>* dynamicTexList  = material.dynamicTexList;
    for (int i= 0; i < dynamicTexList.count; i++) {
        
        if (dynamicTexList[i].isParticleColor) {
            NSLog(@"---");
            [(DynamicTexItem*)dynamicTexList[i]  creatTextureByCurve];
             
        } else {
            /*
            [[ TextureManager default] getTexture:[[Scene_data default]getWorkUrlByFilePath:dynamicTexList[i].url] fun:^(TextureRes * textureRes,DynamicTexListVo* texListVo) {
                DynamicTexItem *dynamicTexItem=(DynamicTexItem*)texListVo;
                dynamicTexItem.textureRes=textureRes;
            } texListVo:dynamicTexList[i]];
            */
               NSLog(@"--url-%@",dynamicTexList[i].url);
              NSLog(@"----");
            [[ TextureManager default]getTexture:[[Scene_data default]getWorkUrlByFilePath:dynamicTexList[i].url] fun:^(NSObject * _Nonnull any) {
                
                NSLog(@"----");
                NSDictionary* bdic=(NSDictionary*)any;
                DynamicTexItem* tempInof= (DynamicTexItem*)bdic[@"info"];
                TextureRes*  ddd=bdic[@"data"];
                tempInof.textureRes=ddd;
                
                
            } wrapType:0 info: dynamicTexList[i] filteType:0 mipmapType:0];
            
            
            
            /*
             TextureManager.getInstance().getTexture(Scene_data.fileRoot + dynamicTexList[i].url, ($textureVo: TextureRes, $texItem: DynamicTexItem) => {
             $texItem.textureRes = $textureVo;
             }, 0, dynamicTexList[i], 0, 1);
             */
        }
        
    }
    
}
/*
public loadDynamicTexUtil(material: MaterialParam): void {
      var dynamicTexList: Array<DynamicTexItem> = material.dynamicTexList;

      for (var i: number = 0; i < dynamicTexList.length; i++) {
          if (dynamicTexList[i].isParticleColor) {
              dynamicTexList[i].creatTextureByCurve();
          } else {
              TextureManager.getInstance().getTexture(Scene_data.fileRoot + dynamicTexList[i].url, ($textureVo: TextureRes, $texItem: DynamicTexItem) => {
                  $texItem.textureRes = $textureVo;
              }, 0, dynamicTexList[i], 0, 1);
          }
      }

  }
*/
-(void)loadMaterial:(Material*)material;
{
    NSMutableArray<TexItem*>* texVec = material.texList;
    for (int i = 0; i < texVec.count; i++) {
        if (texVec[i].isParticleColor || texVec[i].isDynamic || texVec[i].type != 0) {
            continue;
        }
     
    }
}

/*
 private loadMaterial($material: Material): void {
       var texVec: Array<TexItem> = $material.texList;
       for (var i: number = 0; i < texVec.length; i++) {
           if (texVec[i].isParticleColor || texVec[i].isDynamic || texVec[i].type != 0) {
               continue;
           }
           TextureManager.getInstance().getTexture(Scene_data.fileRoot + texVec[i].url, ($textureVo: TextureRes, $texItem: TexItem) => {
               $texItem.textureRes = $textureVo;
           }, texVec[i].wrap, texVec[i], texVec[i].filter, texVec[i].mipmap);
       }
   }
 */

                   
@end

File: test/four/app/src/main/java/com/z3d/material/MaterialManager.java
package com.z3d.material;

import android.graphics.Bitmap;
import android.opengl.GLES20;
import android.opengl.GLUtils;
import android.util.Log;

import com.z3d.base.ByteArray;
import com.z3d.base.ResGC;
import com.z3d.base.Scene_data;
import com.z3d.base.TexTuresBackFun;
import com.z3d.filemodel.TextureManager;
import com.z3d.program.ProgrmaManager;
import com.z3d.program.Shader3D;
import com.z3d.scene.Scene3D;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class MaterialManager extends ResGC {



    public HashMap loadDic;
    public HashMap resDic;

    public MaterialManager(Scene3D val) {
        super(val);
        this.loadDic=new HashMap();
        this.resDic=new HashMap();
    }


    public TextureRes getMaterialByUrl(String url)
    {

        TextureRes textureRes= new TextureRes(scene3D);
        Bitmap b = Bitmap.createBitmap(512, 512, Bitmap.Config.RGB_565);
        textureRes.textTureInt= this.createTexture(b);
        return  textureRes;

    }
    private int createTexture(Bitmap bitmap){
        int[] texture=new int[1];
        if(bitmap!=null&&!bitmap.isRecycled()){
            //生成纹理
            GLES20.glGenTextures(1,texture,0);
            //生成纹理
            GLES20.glBindTexture(GLES20.GL_TEXTURE_2D,texture[0]);
            //设置缩小过滤为使用纹理中坐标最接近的一个像素的颜色作为需要绘制的像素颜色
            GLES20.glTexParameterf(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MIN_FILTER,GLES20.GL_NEAREST);
            //设置放大过滤为使用纹理中坐标最接近的若干个颜色，通过加权平均算法得到需要绘制的像素颜色
            GLES20.glTexParameterf(GLES20.GL_TEXTURE_2D,GLES20.GL_TEXTURE_MAG_FILTER,GLES20.GL_LINEAR);
            //设置环绕方向S，截取纹理坐标到[1/2n,1-1/2n]。将导致永远不会与border融合
            GLES20.glTexParameterf(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_S,GLES20.GL_CLAMP_TO_EDGE);
            //设置环绕方向T，截取纹理坐标到[1/2n,1-1/2n]。将导致永远不会与border融合
            GLES20.glTexParameterf(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_T,GLES20.GL_CLAMP_TO_EDGE);
            //根据以上指定的参数，生成一个2D纹理
            GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, bitmap, 0);
            return texture[0];
        }
        return 0;
    }
    public void addResByte(String url, ByteArray data) {

        if (! this.dic.containsKey(url) && ! this.resDic.containsKey(url)) {
            this.resDic.put(url,data);
        }


    }

 //   -(void)getMaterialByte:(NSString*)url fun:(SuccessMaterial)fun info:(NSDictionary*)info autoReg:(BOOL)autoReg regName:(NSString*)regName shader3DCls:(NSObject*)shader3DCls;

    public void getMaterialByte(String url, MaterialBackFun materialBfun, boolean autoReg, String resName , Shader3D shader3DCls )
    {
        url = url.replace("_byte.txt", ".txt");
        url = url.replace(".txt", "_byte.txt");
        if(this.dic.containsKey(url)){
            materialBfun.Bfun((Material)this.dic.get(url));
            return;
        }
        MaterialLoad materialLoad=new MaterialLoad(materialBfun,null,url,autoReg,resName,shader3DCls);
        if(this.loadDic.containsKey(url)){
            List arr=(List) this.loadDic.get(url);
            arr.add(materialLoad);
            return;
        }
        List newArr=new ArrayList();
        newArr.add(materialLoad);
        this.loadDic.put(url,newArr);
        if(this.resDic.containsKey(url)){
            this.meshByteMaterialByte((ByteArray)this.resDic.get(url), materialLoad);
            this.resDic.remove(url);

        }else{
            Log.d("需要加载功能", "getMaterialByte: ");
        }


    }
    private  void   loadMaterial(Material material)
    {
        List texVec =material.texList;
        for (int i = 0; i < texVec.size(); i++) {

            TexItem texItem=(TexItem)texVec.get(i);
            if (texItem.isParticleColor ||texItem.isDynamic || texItem.type != 0) {
                continue;
            }
           scene3D.textureManager.getTexture(Scene_data.fileRoot+texItem.url, new TexTuresBackFun() {
                @Override
                public void Bfun(TextureRes value) {
                    Log.d("", "Bfun: "+value);
                    texItem.textureRes =value;
                }
            });

        }


    }

    private  void meshByteMaterialByte(ByteArray _byte,MaterialLoad _info)
    {

        Material material = new Material(scene3D);
        material.setByteData(_byte);
        material.url = _info.url;
        this.loadMaterial(material);
        if (_info.autoReg) {
            material.shader = scene3D.progrmaManager.getMaterialProgram(_info.regName,_info.shader3D,material,new ArrayList<Boolean>(),true);
        }
       List arr= (List) this.loadDic.get(_info.url);
       for (int i=0;i<arr.size();i++){
           MaterialLoad materialLoad=(MaterialLoad)arr.get(i);
           materialLoad.fun.Bfun(material);
       }
       this.loadDic.remove(_info.url);
       this.dic.put(_info.url,material);

     }
    public void loadDynamicTexUtil(MaterialParam materialParam) {
        List<DynamicTexItem> dynamicTexList = materialParam.dynamicTexList;
        for (int i = 0; i < dynamicTexList.size(); i++) {
            DynamicTexItem dynamicTexItem=  dynamicTexList.get(i);
            if (dynamicTexItem.isParticleColor) {
                dynamicTexItem.creatTextureByCurve();
            } else {
              scene3D.textureManager.getTexture(Scene_data.fileRoot + dynamicTexItem.url, new TexTuresBackFun() {
                    @Override
                    public void Bfun(TextureRes value) {
                        dynamicTexItem.textureRes = value;
                    }
                });
            }
        }
    }
}
